tests/error_test.c

```diff
@@ -0,0 +1,251 @@
+    int last_seen;
+};
+
+static void error_test_global_cb(int err, void *ctx) {
+    struct error_test_cb_data *cb_data = (struct error_test_cb_data *)ctx;
+    cb_data->global_cb_called = 1;
+    cb_data->last_seen = err;
+}
+
+static void error_test_thread_local_cb(int err, void *ctx) {
+    struct error_test_cb_data *cb_data = (struct error_test_cb_data *)ctx;
+    cb_data->tl_cb_called = 1;
+    cb_data->last_seen = err;
+}
+
+static int error_callback_test_fn(struct aws_allocator *config, void *ctx) {
+
+    struct error_test_cb_data cb_data = {
+            .last_seen = 0,
+            .global_cb_called = 0,
+            .tl_cb_called = 0
+    };
+
+    struct aws_error_info test_error_1 = errors[0];
+    struct aws_error_info test_error_2 = errors[1];
+
+    aws_set_global_error_handler_fn(error_test_global_cb, &cb_data);
+
+    aws_raise_error(test_error_1.error_code);
+
+    ASSERT_NOT_NULL(cb_data.last_seen, "last error should not have been null");
+    ASSERT_TRUE(cb_data.global_cb_called, "Global Callback should have been invoked");
+    ASSERT_FALSE(cb_data.tl_cb_called, "Global Callback should not have been invoked");
+
+    ASSERT_INT_EQUALS(test_error_1.error_code, cb_data.last_seen, "Expected error code %d, but was %d",
+                      test_error_1.error_code, cb_data.last_seen);
+    ASSERT_STR_EQUALS(test_error_1.error_str, aws_error_str(cb_data.last_seen), "Expected error string %s, but got %s",
+                      test_error_1.error_str, aws_error_str(cb_data.last_seen));
+    ASSERT_STR_EQUALS(test_error_1.lib_name, aws_error_lib_name(cb_data.last_seen), "Expected error libname %s, but got %s",
+                      test_error_1.lib_name, aws_error_lib_name(cb_data.last_seen));
+
+    cb_data.last_seen = 0;
+    cb_data.global_cb_called = 0;
+    aws_set_thread_local_error_handler_fn(error_test_thread_local_cb, &cb_data);
+
+    aws_raise_error(test_error_2.error_code);
+    ASSERT_INT_EQUALS(test_error_2.error_code, aws_last_error(), "Expected error code %d, but was %d",
+                      test_error_2.error_code, aws_last_error());
+
+
+    ASSERT_NOT_NULL(cb_data.last_seen, "last error should not have been null");
+    ASSERT_FALSE(cb_data.global_cb_called, "Global Callback should not have been invoked");
+    ASSERT_TRUE(cb_data.tl_cb_called, "Global Callback should have been invoked");
+
+    ASSERT_INT_EQUALS(test_error_2.error_code, cb_data.last_seen, "Expected error code %d, but was %d",
+                      test_error_2.error_code, cb_data.last_seen);
+    ASSERT_STR_EQUALS(test_error_2.error_str, aws_error_str(cb_data.last_seen), "Expected error string %s, but got %s",
+                      test_error_2.error_str, aws_error_str(cb_data.last_seen));
+    ASSERT_STR_EQUALS(test_error_2.lib_name, aws_error_lib_name(cb_data.last_seen), "Expected error libname %s, but got %s",
+                      test_error_2.lib_name, aws_error_lib_name(cb_data.last_seen));
+
+    return 0;
+}
```

Comment:
When we add a thread abstraction we should test the thread-locality of the error APIs.

Reply:
I'll ad a todo.